File: 8-Puzzle/bfs.cpp
#include <iostream>
#include <vector>
#include <chrono>
using namespace std::chrono;
using namespace std;

// Global variables
vector<vector<int>> GOAL_STATE = {{0, 1, 2}, {3, 4, 5}, {6, 7, 8}};

// vector<vector<int>> INITIAL_STATE = {
//      {2,5,0},
//      {1,3,8},
//      {6,4,7}
//  };

vector<vector<int>> INITIAL_STATE = {
     {2,5,8},
     {1,3,0},
     {6,4,7}
 };
//  vector<vector<int>> INITIAL_STATE = {
//      {2,5,8},
//      {1,3,7},
//      {6,4,0}
//  };

// vector<vector<int>> INITIAL_STATE = {
//     {2,5,8},
//     {1,3,7},
//     {6,0,4}
// };

// vector<vector<int>> INITIAL_STATE = {
//     {2,5,8},
//     {1,3,7},
//     {0,6,4}
// };

// vector<vector<int>> INITIAL_STATE = {
//     {2, 5, 8},
//     {0, 3, 7},
//     {1, 6, 4}
// };

struct Node
{
    vector<vector<int>> state;
    Node *parent = nullptr;
    string action;
    int path_cost;

    Node(vector<vector<int>> state, Node *parent, string action, int path_cost)
    {
        this->state = state;
        this->parent = parent;
        this->action = action;
        this->path_cost = path_cost;
    }
};

void printState(vector<vector<int>> state)
{
    for (int i = 0; i < state.size(); i++)
    {
        for (int j = 0; j < state[i].size(); j++)
        {
            cout << state[i][j] << " ";
        }
        cout << endl;
    }
}

vector<vector<int>> moveLeft(vector<vector<int>> state, int row, int col)
{
    // Moves the blank tile to the left
    if (col == 0)
    {
        vector<vector<int>> empty_state = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}};
        return empty_state;
    }
    else
    {
        int temp = state[row][col];
        state[row][col] = state[row][col - 1];
        state[row][col - 1] = temp;
        return state;
    }
}

vector<vector<int>> moveRight(vector<vector<int>> state, int row, int col)
{
    // Moves the blank tile to the right
    if (col == 2)
    {
        vector<vector<int>> empty_state = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}};
        return empty_state;
    }
    else
    {
        int temp = state[row][col];
        state[row][col] = state[row][col + 1];
        state[row][col + 1] = temp;
        return state;
    }
}

vector<vector<int>> moveUp(vector<vector<int>> state, int row, int col)
{
    // Moves the blank tile up
    if (row == 0)
    {
        vector<vector<int>> empty_state = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}};
        return empty_state;
    }
    else
    {
        int temp = state[row][col];
        state[row][col] = state[row - 1][col];
        state[row - 1][col] = temp;
        return state;
    }
}

vector<vector<int>> moveDown(vector<vector<int>> state, int row, int col)
{
    // Moves the blank tile down

    if (row == 2)
    {
        vector<vector<int>> empty_state = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}};
        return empty_state;
    }
    else
    {
        int temp = state[row][col];
        state[row][col] = state[row + 1][col];
        state[row + 1][col] = temp;
        return state;
    }
}

bool isGoalState(vector<vector<int>> state)
{
    // Checks if the current state is the goal state
    for (int i = 0; i < state.size(); i++)
    {
        for (int j = 0; j < state[i].size(); j++)
        {
            if (state[i][j] != GOAL_STATE[i][j])
                return false;
        }
    }
    return true;
}

vector<vector<int>> findIndex(vector<vector<int>> state)
{
    // Finds the index of the blank tile
    vector<vector<int>> index;
    for (int i = 0; i < state.size(); i++)
    {
        for (int j = 0; j < state[i].size(); j++)
        {
            if (state[i][j] == 0)
            {
                index.push_back({i, j});
                return index;
            }
        }
    }
    return {{-1, -1}};
}

vector<Node> expand(Node node)
{
    // Used to compare if action was valid. If invalid, do not add to children vector
    vector<vector<int>> EMPTY = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}};

    // Create resulting vector of children nodes
    vector<Node> children;

    // Find the index of the blank tile and store the resulting row and col
    vector<vector<int>> index = findIndex(node.state);
    int row = index[0][0];
    int col = index[0][1];

    // Create the children nodes
    Node left = Node(moveLeft(node.state, row, col), &node, "left", node.path_cost + 1);
    Node right = Node(moveRight(node.state, row, col), &node, "right", node.path_cost + 1);
    Node up = Node(moveUp(node.state, row, col), &node, "up", node.path_cost + 1);
    Node down = Node(moveDown(node.state, row, col), &node, "down", node.path_cost + 1);

    // Add the children nodes to the children vector
    children.push_back(left);
    children.push_back(right);
    children.push_back(up);
    children.push_back(down);

    // return the vector of children nodes
    return children;
}

Node bfs(vector<vector<int>> initial_state, int *frontier_size, int *reached_size)
{
    // Initialize the starting node
    Node root(initial_state, NULL, "", 0);

    // Check if the initial state is the goal state
    if (isGoalState(root.state))
    {
        *frontier_size = 0;
        *reached_size = 1;
        return root;
    }

    // Initialize the frontier and reached
    vector<Node> frontier;
    frontier.push_back(root);
    vector<Node> reached;

    // Loop until the frontier is empty
    while (frontier.size() != 0)
    {
        // Pop the first element in the frontier
        // Acts as a queue
        Node node = frontier[0];
        frontier.erase(frontier.begin());

        // Store element in reached.
        reached.push_back(node);

        // Check if the current nodes state is the goal state
        if (isGoalState(node.state))
        {
            // If it is the goal state, return the node and the sizes of the frontier and reached
            *reached_size = reached.size();
            *frontier_size = frontier.size();
            return node;
        }

        // Expand the current node to get its children
        vector<Node> children = expand(node);

        // Loop over all current children
        for (int i = 0; i < children.size(); i++)
        {
            // Flag determines if the child already is in reached
            bool flag = false;

            // Account for if one of the children attempted to move the blank tile out of bounds
            if (children[i].state == vector<vector<int>>{{0, 0, 0}, {0, 0, 0}, {0, 0, 0}})
                continue;

            // Loop over all elements in reached to determine if child is already in reached
            for (int j = 0; j < reached.size(); j++)
            {
                if (children[i].state == reached[j].state)
                {
                    flag = true;
                    break;
                }
            }

            // If not in reached, add to reached.
            // Also add to frontier to expand later
            if (!flag)
            {
                frontier.push_back(children[i]);
                reached.push_back(children[i]);
            }
        }
    }

    // If the frontier is empty and no solution has been found, return the initial node.
    return root;
}

int main()
{
    int frontier_size = 0, reached_size = 0;

    // If no empty space then there is not solution
    if (findIndex(INITIAL_STATE)[0][0] == -1)
    {
        cout << "Invalid initial state" << endl;
        return 0;
    }

    // Print the initial state
    printState(INITIAL_STATE);
    cout << endl;

    // Time how long it takes for the algorithm to complete
    auto start = high_resolution_clock::now();
    Node end = bfs(INITIAL_STATE, &frontier_size, &reached_size);
    auto stop = high_resolution_clock::now();

    // If the initial state is the goal state, no solution was found
    if (end.state == INITIAL_STATE)
    {
        cout << "No solution found" << endl;
        return 0;
    }

    // If solution was found, print the solution and other information
    printState(end.state);
    cout << "\nNumber of steps: " << end.path_cost << endl;
    cout << "frontier size: " << frontier_size << endl;
    cout << "reached size: " << reached_size << endl;
    cout << "Time taken: " << duration_cast<milliseconds>(stop - start).count() << " milliseconds" << endl;

    return 0;
}

File: 8-Puzzle/bfs.py
from collections import deque
import numpy as np
import time
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning) 

GOAL_STATE = [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
GOAL_STATE = np.array(GOAL_STATE)

class Node:
    def __init__(self, state, parent, action, path_cost):
        self.state = state
        self.parent = parent
        self.action = action
        self.path_cost = path_cost


def printState(state):
    """
    Prints the board.
    Requires a 3x3 list
    """
    for i in range(3):
        for j in range(3):
            print(state[i][j], end = " ")
        print()
    print()


def isGoalState(state):
    """
    Checks if the state is the goal state.
    """
    if np.array_equal(state, GOAL_STATE):
        return True
    else:
        return False


def getIndex(state):
    """
    Returns the index (i and j) of the blank tile
    """
    empty = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    if np.array_equal(state, empty):
        return -1, -1
    for i in range(3):
        for j in range(3):
            if state[i][j] == 0:
                return i, j


def moveUp(state):
    """
    Moves the blank tile up
    """
    row, col = getIndex(state)

    if row == 0 or row == -1 and col == -1:
        return np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    else:
        new_state = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
        for i in range(3):
            for j in range(3):
                new_state[i][j] = state[i][j]

        new_state[row][col], new_state[row-1][col] = new_state[row-1][col], new_state[row][col]
        new_state = np.array(new_state)
        return new_state

def moveDown(state):
    """
    Moves the blank tile down
    """
    row, col = getIndex(state)

    if row == 2 or row == -1 and col == -1:
        return np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    else:
        new_state = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
        for i in range(3):
            for j in range(3):
                new_state[i][j] = state[i][j]

        new_state[row][col], new_state[row+1][col] = new_state[row+1][col], new_state[row][col]
        new_state = np.array(new_state)
        return new_state

def moveLeft(state):
    """
    Moves the blank tile left
    """
    row, col = getIndex(state)

    if col == 0 or row == -1 and col == -1:
        return np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    else:
        new_state = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
        for i in range(3):
            for j in range(3):
                new_state[i][j] = state[i][j]

        new_state[row][col], new_state[row][col-1] = new_state[row][col-1], new_state[row][col]
        new_state = np.array(new_state)
        return new_state

def moveRight(state):
    """
    Moves the blank tile right
    """
    row, col = getIndex(state)

    if col == 2 or row == -1 and col == -1:
        return np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    else:
        new_state = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
        for i in range(3):
            for j in range(3):
                new_state[i][j] = state[i][j]

        new_state[row][col], new_state[row][col+1] = new_state[row][col+1], new_state[row][col]
        new_state = np.array(new_state)
        return new_state

def breadthFirstSearch(initial_node):
    """
    Takes an initial state. Will expand the initial state.
    Priortizes expanding the states at the same depth using
    a FIFO queue. If one of the children is the goal state,
    then that child is returned. If no solution is found,
    then None is returned.
    """
    initial_state = initial_node.state
    initial_state = np.array(initial_state)
    empty = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    if isGoalState(initial_state):
        return initial_node, 0, 1
    frontier = deque([initial_node])
    reached = np.array([initial_node])
    while frontier:
        node = frontier.popleft()
        child = expand(node)
        for i in range(len(child)):
            state = child[i].state
           
           
            state = np.array(state)
            if np.array_equal(state, empty):
                continue
            if isGoalState(state):
                return child[i], len(frontier), len(reached)
            
            # Check if the current state node has already been reached
            flag = False
            for j in range(len(reached)):
                if np.array_equal(state, reached[j].state):
                    flag = True
                    break

            if not flag:
                frontier.append(child[i])
                reached = np.append(reached, child[i])

    return None, len(frontier), len(reached)


def expand(node):
    """
    Performs the four possible moves. If a move is not possible,
    the move function returns None. If that happens, it will
    be ignored.
    """
    state = node.state
    expanded_nodes = []

    expanded_nodes.append(
        Node(moveUp(state), node, "Up", node.path_cost+1))
    expanded_nodes.append(
        Node(moveDown(state), node, "Down", node.path_cost+1))
    expanded_nodes.append(
        Node(moveLeft(state), node, "Left", node.path_cost+1))
    expanded_nodes.append(
        Node(moveRight(state), node, "Right", node.path_cost+1))

    return expanded_nodes

def allActions(node):
    """
    Prints actions and states from the initial state to the goal state.
    """
    actions = []
    states = []

    while node.parent != None:
        actions.append(node.action)
        states.append(node.state)
        node = node.parent
    actions.reverse()
    states.reverse()
    for i in range(len(actions)):
        print(f"Step {i+1}:")
        print(f"Action: {actions[i]}")
        print("\nNew State:")
        printState(states[i])
    print()

def menu():
    print("8-Puzzle Solver")
    print("---------------")
    print("To Note: ")
    print("1. The blank tile is represented by a 0.\n")
    print("2. The goal state is:")
    printState(GOAL_STATE)
    print()

def main():
    menu()

    # Takes 10 steps
    # initial_state = [
    #     [2, 5, 0], 
    #     [1, 3, 8], 
    #     [6, 4, 7]
    # ]

    # Takes 11 steps and much longer than the one above.
    initial_state = [
        [2, 5, 8], 
        [1, 3, 0], 
        [6, 4, 7]
    ]

    # Apprently complicated
    # initial_state = [
    #     [8, 1, 2], 
    #     [7, 0, 3], 
    #     [6, 5, 4]
    # ]
    initial_state = np.array(initial_state)

    initial_node = Node(initial_state, None, None, 0)

    print("Initial State:")
    printState(initial_state)

    start = time.time()
    goal_node, len_of_frontier, len_of_reached = breadthFirstSearch(initial_node)
    end = time.time()

    empty = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    if goal_node == None:
        print("No solution found")
    else:
        allActions(goal_node)
        print("Total Number of Steps: ", goal_node.path_cost)
        print("Length of Frontier: ", len_of_frontier)
        print("Length of Reached: ", len_of_reached)
    print(f"Time: {end-start} seconds")


if __name__ == '__main__':
    main()
